shoes-core/lib/shoes/http_wrapper.rb

```diff
@@ -0,0 +1,55 @@
+require 'uri'
+require 'net/http'
+
+# Wrapper for Net::HTTP to provide us to the functionality that we need which
+# isn't easily available from any particular gem out there :(
+#
+# Specific things this class does that we need include:
+#
+#   1) Reading in chunks to trigger progress
+#   2) Allow for graceful redirection
+#
+# Most of the gems out there do 2) but not 1), or do both but are untested on
+# JRuby.
+#
+class Shoes
+  class HttpWrapper
+    def execute(url, meth, body, headers = {}, redirects_left = 5, &blk)
+      uri = URI(url)
```

Comment:
Ah, also just realized one of the nice bits about `URI()`--if you hand it something that's a `URI` instance already, it quietly hands it along. `URI.parse` raises if the thing isn't a string to parse.

Still cool with the change, but I think that's the idea behind those conversion methods.